packages/astro/src/core/build/plugins/plugin-css.ts

```diff
@@ -25,236 +25,312 @@ interface PluginOptions {
 	target: 'client' | 'server';
 }
 
-export function rollupPluginAstroBuildCSS(options: PluginOptions): VitePlugin[] {
+/***** ASTRO PLUGIN *****/
+
+export function pluginCSS(
+	options: StaticBuildOptions,
+	internals: BuildInternals
+): AstroBuildPlugin {
+	return {
+		build: 'both',
+		hooks: {
+			'build:before': ({ build }) => {
+				let plugins = rollupPluginAstroBuildCSS({
+					buildOptions: options,
+					internals,
+					target: build === 'ssr' ? 'server' : 'client',
+				});
+
+				return {
+					vitePlugin: plugins,
+				};
+			},
+		},
+	};
+}
+
+/***** ROLLUP SUB-PLUGINS *****/
+
+function rollupPluginAstroBuildCSS(options: PluginOptions): VitePlugin[] {
 	const { internals, buildOptions } = options;
 	const { settings } = buildOptions;
 
 	let resolvedConfig: ResolvedConfig;
 
-	function createNameHash(baseId: string, hashIds: string[]): string {
-		const baseName = baseId ? npath.parse(baseId).name : 'index';
-		const hash = crypto.createHash('sha256');
-		for (const id of hashIds) {
-			hash.update(id, 'utf-8');
-		}
-		const h = hash.digest('hex').slice(0, 8);
-		const proposedName = baseName + '.' + h;
-		return proposedName;
-	}
+	// stylesheet filenames are kept in here until "post", when they are rendered and ready to be inlined
+	const pagesToCss = new Map<string, Map<string, { order: number; depth: number }>>();
+	const pagesToPropagatedCss = new Map<string, Map<string, Set<string>>>();
 
```

Comment:
nit: I personally find keeping a temporary map instead of updating the `pageData` directly a bit hard to follow. Is it necessary to refactor this for this feature to work?

I don't have a strong opinion of going back to the previous pattern since the work is already done, but seeing the `.set().get()` is causing an itch in me 😬